File: tests/typesafety/python/exercise_money_solution.py
# expect: passes
# title: The type system separates the currencies
# improves: exercise_money_start
"""One class per currency, and `add` constrained to a single one of them.
Cents are integers, so the totals are exact."""

from dataclasses import dataclass, replace


@dataclass(frozen=True)
class Usd:
    cents: int


@dataclass(frozen=True)
class Eur:
    cents: int


def add[M: (Usd, Eur)](a: M, b: M) -> M:
    return replace(a, cents=a.cents + b.cents)


def basket_total() -> Usd:
    return add(Usd(1999), Usd(500))

File: tests/typesafety/typescript/exercise_money_solution.ts
// expect: passes

type Currency = "USD" | "EUR";

type Money<C extends Currency> = {
  readonly currency: C;
  readonly cents: number;
};

function money<C extends Currency>(currency: C, cents: number): Money<C> {
  return { currency, cents };
}

// NoInfer fixes C from the first amount, so the second cannot widen it to USD | EUR.
function add<C extends Currency>(a: Money<C>, b: Money<NoInfer<C>>): Money<C> {
  return { currency: a.currency, cents: a.cents + b.cents };
}

export function basketTotal(): Money<"USD"> {
  return add(money("USD", 1999), money("USD", 500));
}
